File: dotnet/src/HybridRow/Schemas/TypeKind.cs
// ------------------------------------------------------------
//  Copyright (c) Microsoft Corporation.  All rights reserved.
// ------------------------------------------------------------

namespace Microsoft.Azure.Cosmos.Serialization.HybridRow.Schemas
{
    using System.Runtime.Serialization;
    using Newtonsoft.Json;
    using Newtonsoft.Json.Converters;

    /// <summary>Describes the logical type of a property.</summary>
    [JsonConverter(typeof(StringEnumConverter), true)] // camelCase:true
    public enum TypeKind
    {
        /// <summary>Reserved.</summary>
        Invalid = 0,

        /// <summary>The literal null.</summary>
        /// <remarks>
        /// When used as a fixed column, only a presence bit is allocated.  When used as a sparse
        /// column, a sparse value with 0-length payload is written.
        /// </remarks>
        Null,

        /// <summary>
        /// A boolean property.  Boolean properties are allocated a single bit when schematized within
        /// a row.
        /// </summary>
        [EnumMember(Value = "bool")]
        Boolean,

        /// <summary>8-bit signed integer.</summary>
        Int8,

        /// <summary>
        /// 16-bit signed integer.
        /// </summary>
        Int16,

        /// <summary>32-bit signed integer.</summary>
        Int32,

        /// <summary>64-bit signed integer.</summary>
        Int64,

        /// <summary>8-bit unsigned integer.</summary>
        [EnumMember(Value = "uint8")]
        UInt8,

        /// <summary>16-bit unsigned integer.</summary>
        [EnumMember(Value = "uint16")]
        UInt16,

        /// <summary>32-bit unsigned integer.</summary>
        [EnumMember(Value = "uint32")]
        UInt32,

        /// <summary>64-bit unsigned integer.</summary>
        [EnumMember(Value = "uint64")]
        UInt64,

        /// <summary>Variable length encoded signed integer.</summary>
        [EnumMember(Value = "varint")]
        VarInt,

        /// <summary>Variable length encoded unsigned integer.</summary>
        [EnumMember(Value = "varuint")]
        VarUInt,

        /// <summary>32-bit IEEE 754 floating point value.</summary>
        Float32,

        /// <summary>64-bit IEEE 754 floating point value.</summary>
        Float64,

        /// <summary>128-bit IEEE 754-2008 floating point value.</summary>
        Float128,

        /// <summary>128-bit floating point value.  See <see cref="decimal" /></summary>
        Decimal,

        /// <summary>
        /// 64-bit date/time value in 100ns increments from C# epoch.  See <see cref="System.DateTime" />
        /// </summary>
        [EnumMember(Value = "datetime")]
        DateTime,

        /// <summary>
        /// 64-bit date/time value in milliseconds increments from Unix epoch.  See <see cref="UnixDateTime" />
        /// </summary>
        [EnumMember(Value = "unixdatetime")]
        UnixDateTime,

        /// <summary>128-bit globally unique identifier (in little-endian byte order).</summary>
        Guid,

        /// <summary>12-byte MongoDB Object Identifier (in little-endian byte order).</summary>
        [EnumMember(Value = "mongodbobjectid")]
        MongoDbObjectId,

        /// <summary>Zero to MAX_ROW_SIZE bytes encoded as UTF-8 code points.</summary>
        Utf8,

        /// <summary>Zero to MAX_ROW_SIZE untyped bytes.</summary>
        Binary,

        /// <summary>An object property.</summary>
        Object,

        /// <summary>An array property, either typed or untyped.</summary>
        Array,

        /// <summary>A set property, either typed or untyped.</summary>
        Set,

        /// <summary>A map property, either typed or untyped.</summary>
        Map,

        /// <summary>A tuple property.  Tuples are typed, finite, ordered, sets.</summary>
        Tuple,

        /// <summary>A tagged property.  Tagged properties pair one or more typed values with an API-specific uint8 type code.</summary>
        Tagged,

        /// <summary>A row with schema.</summary>
        /// <remarks>May define either a top-level table schema or a UDT (nested row).</remarks>
        Schema,

        /// <summary>An untyped sparse field.</summary>
        /// <remarks>May only be used to define the type within a nested scope (e.g. <see cref="Object"/> or <see cref="Array"/>.</remarks>
        Any,
    }
}

File: java/src/main/java/com/azure/data/cosmos/serialization/hybridrow/schemas/TypeKind.java
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

package com.azure.data.cosmos.serialization.hybridrow.schemas;

// TODO: DANOBLE: Fixup JSON-serialized naming for agreement with the dotnet code

import it.unimi.dsi.fastutil.ints.Int2ObjectMap;
import it.unimi.dsi.fastutil.ints.Int2ObjectOpenHashMap;

/**
 * Describes the logical type of a property.
 */
public enum TypeKind {
    /**
     * Reserved.
     */
    INVALID(0),

    /**
     * The literal null.
     * <p>
     * When used as a fixed column, only a presence bit is allocated.  When used as a sparse column, a sparse value with
     * 0-length payload is written.
     */
    NULL(1),

    /**
     * A boolean property.
     * <p>
     * Boolean properties are allocated a single bit when schematized within a row.
     */
    BOOLEAN(2),

    /**
     * 8-bit signed integer.
     */
    INT_8(3),

    /**
     * 16-bit signed integer.
     */
    INT_16(4),

    /**
     * 32-bit signed integer.
     */
    INT_32(5),

    /**
     * 64-bit signed integer.
     */
    INT_64(6),

    /**
     * 8-bit unsigned integer.
     */
    UINT_8(7),

    /**
     * 16-bit unsigned integer.
     */
    UINT_16(8),

    /**
     * 32-bit unsigned integer.
     */
    UINT_32(9),

    /**
     * 64-bit unsigned integer.
     */
    UINT_64(10),

    /**
     * Variable length encoded signed integer.
     */
    VAR_INT(11),

    /**
     * Variable length encoded unsigned integer.
     */
    VAR_UINT(12),

    /**
     * 32-bit IEEE 754 floating point value.
     */
    FLOAT_32(13),

    /**
     * 64-bit IEEE 754 floating point value.
     */
    FLOAT_64(14),

    /**
     * 128-bit IEEE 754-2008 floating point value.
     */
    FLOAT_128(15),

    /**
     * 128-bit floating point value.
     *
     * @see java.math.BigDecimal
     */
    DECIMAL(16),

    /**
     * 64-bit date/time value in 100ns increments from C# epoch.
     *
     * @see java.time.OffsetDateTime
     */
    DATE_TIME(17),

    /**
     * 64-bit date/time value in milliseconds increments from Unix epoch.
     *
     * @see com.azure.data.cosmos.serialization.hybridrow.UnixDateTime
     */
    UNIX_DATE_TIME(18),

    /**
     * 128-bit globally unique identifier (in little-endian byte order).
     */
    GUID(19),

    /**
     * 12-byte MongoDB Object Identifier (in little-endian byte order).
     */
    MONGODB_OBJECT_ID(20),

    /**
     * Zero to MAX_ROW_SIZE bytes encoded as UTF-8 code points.
     */
    UTF_8(21),

    /**
     * Zero to MAX_ROW_SIZE untyped bytes.
     */
    BINARY(22),

    /**
     * An object property.
     */
    OBJECT(23),

    /**
     * An array property, either typed or untyped.
     */
    ARRAY(24),

    /**
     * A set property, either typed or untyped.
     */
    SET(25),

    /**
     * A map property, either typed or untyped.
     */
    MAP(26),

    /**
     * A tuple property.  Tuples are typed, finite, ordered, sets.
     */
    TUPLE(27),

    /**
     * A tagged property.
     * <p>
     * Tagged properties pair one or more typed values with an API-specific uint8 type code.
     */
    TAGGED(28),

    /**
     * A row with schema.
     * <p>
     * May define either a top-level table schema or a UDT (nested row).
     */
    SCHEMA(29),

    /**
     * An untyped sparse field.
     * <p>
     * May only be used to define the type within a nested scope.
     */
    ANY(30);

    public static final int BYTES = Integer.BYTES;
    private static Int2ObjectMap<TypeKind> mappings;
    private int value;

    TypeKind(int value) {
        this.value = value;
        mappings().put(value, this);
    }

    public int value() {
        return this.value;
    }

    public static TypeKind from(int value) {
        return mappings().get(value);
    }

    private static Int2ObjectMap<TypeKind> mappings() {
        if (mappings == null) {
            synchronized (TypeKind.class) {
                if (mappings == null) {
                    mappings = new Int2ObjectOpenHashMap<>();
                }
            }
        }
        return mappings;
    }
}
